core/nightshade/src/nightshade_task.rs

```diff
@@ -216,32 +227,39 @@ impl<P: Send + Debug + Clone + Serialize + 'static> NightshadeTask<P> {
         // Get author of the payload inside this message
         let author = message.state.bare_state.endorses.author;
 
-        // Check we already have such payload and request it otherwise
-        if let Some(signed_block) = &self.authority_blocks[author] {
-            if signed_block.block.hash() == message.state.block_hash() {
-                if let Err(e) =
+        // Check if this proposal was confirmed
+        if self.confirmed_proposals[author] {
+            if let Some(signed_proposal) = &self.proposals[author] {
+                if signed_proposal.block_proposal.hash == message.state.block_hash() {
+                    if let Err(e) =
                     self.nightshade_as_mut_ref().update_state(message.sender_id, message.state)
-                {
-                    warn!(target: "nightshade", "{}", e);
+                    {
+                        warn!(target: "nightshade", "{}", e);
+                    }
+                } else {
+                    self.nightshade_as_mut_ref().set_adversary(author);
                 }
             } else {
-                self.nightshade_as_mut_ref().set_adversary(author);
+                panic!("Signed proposal must exist if it was already confirmed.");
             }
         } else {
-            // TODO: This message is discarded if we haven't receive the payload yet. We can store it
-            // in a queue instead, and process it after we have the payload.
-
-            let gossip = Gossip::new(
-                self.nightshade.as_ref().unwrap().owner_id,
-                author,
-                GossipBody::PayloadRequest(vec![author]),
-                self.owner_secret_key.as_ref().unwrap(),
+            // TODO: This message is discarded if we haven't received the proposal yet.
+            tokio::spawn(
```

Comment:
@ilblackdragon we should send payload request gossip through the network, rather than requesting it to the mempool (as it was before). In this point, we don't have the signed hash at this point, and haven't collected the proposal, so we should not request it to the mempool yet.

Reply:
btw, I found this, because NS was receiving blocks confirmation from payloads that it didn't have at the moment.